Versioning refactor; take 2, part 2

LukeShu opened the pull request:
## Description

This PR is the continuation of #4003; which was me taking over @aidanhahn's #3993.  The big goal of this PR is picking up all the goodness from Aidan's original PR that I dropped for #4003. That is: Doing cleanup enabled by the changes in #4003.

## Related Issues
- #3993
- #4003

## Testing
I've used this to drive a v2.1.2 release in the https://github.com/aidanhahn/emissary repo.

## Checklist

<!--
  Please review the requirements for each checkbox, and check them
  off (change "[ ]" to "[x]") as you verify that they are complete.
-->

 - [x] I made sure to update `CHANGELOG.md`. - no applicable changes
   
   Remember, the CHANGELOG needs to mention:
    + Any new features
    + Any changes to our included version of Envoy
    + Any non-backward-compatible changes
    + Any deprecations
 
 - [x] This is unlikely to impact how Ambassador performs at scale. - minimal runtime changes
 
   Remember, things that might have an impact at scale include:
    + Any significant changes in memory use that might require adjusting the memory limits
    + Any significant changes in CPU use that might require adjusting the CPU limits
    + Anything that might change how many replicas users should use
    + Changes that impact data-plane latency/scalability
 
 - [x] My change is adequately tested. - I'd say so
 
   Remember when considering testing:
    + Your change needs to be specifically covered by tests.
       + Tests need to cover all the states where your change is relevant: for example, if you add a behavior that can be enabled or disabled, you'll need tests that cover the enabled case and tests that cover the disabled case. It's not sufficient just to test with the behavior enabled.
    + You also need to make sure that the _entire area being changed_ has adequate test coverage.
       + If existing tests don't actually cover the entire area being changed, add tests.
       + This applies even for aspects of the area that you're not changing – check the test coverage, and improve it if needed!
    + We should lean on the bulk of code being covered by unit tests, but...
    + ... an end-to-end test should cover the integration points
 
 - [x] I updated `DEVELOPING.md` with any any special dev tricks I had to use to work on this code efficiently. - no tricks

kflynn commented on charts/charts.mk:
```diff
@@ -1,97 +1,66 @@
+	{ $(OSS_HOME)/venv/bin/python $(OSS_HOME)/charts/scripts/update_chart_image_values.py \
+	  --values-file=$(EMISSARY_CHART)/values.yaml \
+	  --tag=$(IMAGE_TAG) \
+	  --repo=$(IMAGE_REPO); }
```
What's the value in the `{}` here? Just visual cues for what the command really is?

LukeShu replied:
So it'll tell me what the mistake is if I forget a `\` at the end of a line.

kflynn commented on cmd/busyambassador/main.go:
```diff
@@ -6,110 +6,62 @@
-			line := scanner.Text()
-
+	// Keep this parsing logic in-sync with VERSION.py.
+	version := "dirty"
+	if verBytes, err := os.ReadFile("/buildroot/ambassador/python/ambassador.version"); err == nil {
```
This needs a comment. We _check_ for `err == nil`, but if `err` is not `nil` we don't log or anything, we just silently set `version` to `dirty`. The old version talked at length about why we don't log, and (obviously) I thought that was important.

kflynn commented on python/ambassador/VERSION.py:
```diff
@@ -13,68 +13,20 @@
 import os
-from typing import NamedTuple
+
+Version = "MISSING(FILE)"
+Commit = "MISSING(FILE)"
```
Same note about comments here... even though (my fault, I'm sure) they weren't there before. 🙁

kflynn commented on python/ambassador/VERSION.py:
```diff
@@ -13,68 +13,20 @@
-        print("git.commit:      %s" % Build.git.commit)
-        print("git.dirty:       %s" % Build.git.dirty)
-        print("git.description: %s" % Build.git.description)
-    else:   # compact
-        print("%s (%s at %s on %s%s)" %
-              (Version, Build.git.description, Build.git.commit, Build.git.branch,
-               " - dirty" if Build.git.dirty else ""))
```
I've actually been known to use this...

kflynn replied:
These days you can also `busyambassador version`... it's still a bit different, but it's not a big deal.

kflynn commented on builder/builder.mk:
```diff
@@ -333,10 +333,10 @@ push-dev: docker/$(LCNAME).docker.tag.local
 		fi; \
 		if [[ '$(VERSION)' != *-* ]]; then \
-			printf "$(RED)push-dev: BUILD_VERSION $(BUILD_VERSION) is not a dev version$(END)\n" >&2 ;\
+			printf "$(RED)push-dev: BUILD_VERSION $(patsubst v%,%,$(VERSION)) is not a dev version$(END)\n" >&2 ;\
```
Could this just be

```
VERSION_NO_V=$(patsubst v%,%,$(VERSION))
```

and then use $(VERSION_NO_V) everywhere instead of repeating the `patsubst`?

LukeShu replied:
Hmm, the "begone with BUILD_VERSION" commit should have changed this to
```
printf "$(RED)push-dev: VERSION $(VERSION) is not a dev version$(END)\n" >&2 ;\
```

kflynn replied:
Yeah, but did _all_ those `patsubst` calls go away in `begone with BUILD_VERSION`?

LukeShu replied:
Most of them did.  And the ones that remain should be painful and verbose, because they are indicators of remaining inconsistency in the build system.

kflynn replied:
👍

kflynn commented on builder/builder.mk:
```diff
@@ -759,7 +756,7 @@ release/promote-oss/dev-to-rc:
 		printf "$(CYN)==> $(GRN)found version $(BLU)$$dev_version$(GRN) for $(BLU)$$commit$(GRN) in S3...$(END)\n" ;\
 		veroverride=$(patsubst v%,%,$(VERSION)) ; \
-		tag=$$(echo $(patsubst v%,%,$(VERSION)) | tr '+' '-') ; \
+		tag=$$veroverride ; \
```
The `tr '+' '-'` isn't still important here? IIRC it was there in the first place because of Docker tags...

LukeShu replied:
`goversion` shouldn't ever spit out a version with a `+` in it.

kflynn replied:
👍

kflynn commented on builder/builder.sh:
```diff
@@ -289,39 +289,6 @@ bootstrap() {
     fi
 }
 
-module_version() {
```
MAN it's nice to see this go. 😁

kflynn reviewed (changes requested):
This is gonna be great. 🙂  I do have some questions though...

LukeShu commented:
v2:
 - add info to the commit message of the `builder.mk, release.mk: s/\$\((BUILD|RELEASE)_VERSION\)/$(patsubst v%,%,$(VERSION))/g` commit
 - the following changes:

```patch
diff --git a/cmd/busyambassador/main.go b/cmd/busyambassador/main.go
index 5c90ff154..d4f931782 100644
--- a/cmd/busyambassador/main.go
+++ b/cmd/busyambassador/main.go
@@ -40,7 +40,12 @@ func main() {
 	// recompiling anything.
 	//
 	// Keep this parsing logic in-sync with VERSION.py.
-	version := "dirty"
+	//
+	// We don't report or log errors here, we just silently fall back to a static "MISSING(XXX)"
+	// string.  This is in-part because the code in main() here is running _wicked_ early, and
+	// logging setup hasn't happened yet.  Also because any errors will be evident when the
+	// version number gets logged and it's this static string.
+	version := "MISSING(FILE)"
 	if verBytes, err := os.ReadFile("/buildroot/ambassador/python/ambassador.version"); err == nil {
 		verLines := strings.Split(string(verBytes), "\n")
 		for len(verLines) < 2 {
diff --git a/python/ambassador/VERSION.py b/python/ambassador/VERSION.py
index 2b6fc3067..496aa999e 100644
--- a/python/ambassador/VERSION.py
+++ b/python/ambassador/VERSION.py
@@ -14,11 +14,15 @@
 
 import os
 
+# Keep this in-sync with cmd/busyambassador/main.go.
+#
+# We don't report or log errors here, we just silently fall back to some static "MISSING(XXX)"
+# strings.  This is in-part because the code here is running pretty early, and logging setup hasn't
+# happened yet.  Also because any errors will be evident when the version number gets logged and
+# it's this static string.
 Version = "MISSING(FILE)"
 Commit = "MISSING(FILE)"
-
 try:
-    # Keep this in-sync with cmd/busyambassador/main.go.
     with open(os.path.join(os.path.dirname(__file__), "..", "ambassador.version")) as version:
         info = version.read().split('\n')
         while len(info) < 2:
@@ -26,7 +30,5 @@ try:
 
         Version = info[0]
         Commit = info[1]
-
-
 except FileNotFoundError:
     pass
```

LukeShu commented:
v3:
```patch
diff --git a/.github/workflows/publish-manifests.yml b/.github/workflows/publish-manifests.yml
index eecb0accf..29b0f8e26 100644
--- a/.github/workflows/publish-manifests.yml
+++ b/.github/workflows/publish-manifests.yml
@@ -38,7 +38,7 @@ jobs:
           SLACK_WEBHOOK_URL: ${{ secrets.SLACK_WEBHOOK_URL }}
         with:
           status: ${{ job.status }}
-          success_text: 'Manifests successfully published for ${steps.step-publish-yaml.published_version}'
+          success_text: 'Manifests successfully published for ${{steps.step-publish-yaml.published_version}}'
           failure_text: '${env.GITHUB_WORKFLOW} (${env.GITHUB_RUN_NUMBER}) build failed'
           cancelled_text: '${env.GITHUB_WORKFLOW} (${env.GITHUB_RUN_NUMBER}) build was cancelled'
           fields: |
```

LukeShu commented:
v4:
```patch
diff --git a/releng/chart-create-gh-release b/releng/chart-create-gh-release
index 2893314e6..3d3442316 100755
--- a/releng/chart-create-gh-release
+++ b/releng/chart-create-gh-release
@@ -14,7 +14,7 @@ def main() -> int:
         sys.stderr.write(f'Usage: CHART_VERSION=v7.Y.Z {os.path.basename(sys.argv[0])}\n')
         return 2
 
-    content_gittag = "chart/{chart_version}"
+    content_gittag = f"chart/{chart_version}"
 
     content_title = f"Emissary Ingress Chart {chart_version[1:]}"
 
@@ -28,7 +28,6 @@ View changelog - https://github.com/emissary-ingress/emissary/blob/master/charts
 """
     in_changelog = False
     for line in fileinput.FileInput("charts/emissary-ingress/CHANGELOG.md"):
-        print(f"line={repr(line)}")
         if in_changelog:
             if line.startswith("## v"):
                 break
```

LukeShu commented:
v5:
```patch
diff --git a/builder/builder.mk b/builder/builder.mk
index 8e19b49aa..29bef0ff7 100644
--- a/builder/builder.mk
+++ b/builder/builder.mk
@@ -823,8 +823,8 @@ release/promote-oss/to-hotfix:
 			exit 1 ;\
 		fi ;\
 		[[ "$$hotfix_tag" =~ ^[0-9]+\.[0-9]+\.[0-9]+-hf\.[0-9]+\+[0-9]+$$ ]] || (printf '$(RED)ERROR: tag %s does not look like a hotfix tag\n' "$$hotfix_tag"; exit 1) ;\
-		$(OSS_HOME)/releng/release-wait-for-commit --commit $$HOTFIX_COMMIT --s3-key passed-pr ;\
-		dev_version=$$(aws s3 cp s3://$(AWS_S3_BUCKET)/passed-pr/$$HOTFIX_COMMIT -) ;\
+		$(OSS_HOME)/releng/release-wait-for-commit --commit $$HOTFIX_COMMIT --s3-key dev-builds ;\
+		dev_version=$$(aws s3 cp s3://$(AWS_S3_BUCKET)/dev-builds/$$HOTFIX_COMMIT -) ;\
 		if [ -z "$$dev_version" ]; then \
 			printf "$(RED)==> found no passed dev version for $$HOTFIX_COMMIT in S3...$(END)\n" ;\
 			exit 1 ;\
@@ -858,8 +858,8 @@ release/promote-oss/to-ga:
 	@[[ "$(VERSION)" =~ ^v[0-9]+\.[0-9]+\.[0-9]+(-ea)?$$ ]] || (printf '$(RED)ERROR: VERSION=%s does not look like a GA tag\n' "$(VERSION)"; exit 1)
 	@set -e; { \
 	  commit=$$(git rev-parse HEAD) ;\
-	  $(OSS_HOME)/releng/release-wait-for-commit --commit $$commit --s3-key passed-builds ; \
-	  dev_version=$$(aws s3 cp s3://$(AWS_S3_BUCKET)/passed-builds/$$commit -) ;\
+	  $(OSS_HOME)/releng/release-wait-for-commit --commit $$commit --s3-key dev-builds ; \
+	  dev_version=$$(aws s3 cp s3://$(AWS_S3_BUCKET)/dev-builds/$$commit -) ;\
 	  if [ -z "$$dev_version" ]; then \
 	    printf "$(RED)==> found no passed dev version for $$commit in S3...$(END)\n" ;\
 	    exit 1 ;\
diff --git a/charts/scripts/check_updated_changelog.sh b/charts/scripts/check_updated_changelog.sh
index 9f425687c..9bde6dbee 100755
--- a/charts/scripts/check_updated_changelog.sh
+++ b/charts/scripts/check_updated_changelog.sh
@@ -17,7 +17,7 @@ CHART_DIR="${TOP_DIR}/${CHART_NAME}"
 if [[ ${CHART_VERSION:-} != v7.* ]]; then
     abort "CHART_VERSION must be set to a 'v7.*' string"
 fi
-chart_version=${CHART_VERSION%v}
+chart_version=${CHART_VERSION#v}
 if [[ $chart_version = *-rc.* ]] ; then
     chart_version=${chart_version%-*}
 fi
diff --git a/charts/scripts/common.sh b/charts/scripts/common.sh
index 270f8de69..846498864 100755
--- a/charts/scripts/common.sh
+++ b/charts/scripts/common.sh
@@ -134,38 +134,3 @@ cleanup () {
 
   rm -rf "$VALUES_DIR"
 }
-
-create_chart_release() {
-    if [[ -z "${GH_TOKEN}" ]] ; then
-        gh auth login
-    fi
-    tag="chart/v${1}"
-    chart_version=${1}
-    chart_dir=$2
-    export chart_version_no_v=$1
-    title=`envsubst < ${chart_dir}/RELEASE_TITLE.tpl`
-    repo_full_name="emissary-ingress/emissary"
-    token="${GH_RELEASE_TOKEN}"
-    description=`envsubst < ${chart_dir}/RELEASE.tpl`
-    description=`envsubst < ${chart_dir}/RELEASE.tpl | awk '{printf "%s\\n", $0}'`
-    mkdir -p /tmp
-    printf "${description}" > /tmp/CHANGELOG.md
-    echo '' >> /tmp/CHANGELOG.md
-    in_changelog=false
-    while IFS= read -r line ; do
-        if [[ ${in_changelog} == true ]] ; then
-            if [[ "${line}" =~ "## v" ]] ; then
-                break
-            fi
-            if [[ -n "${line}" ]] ; then
-                echo ${line} >> /tmp/CHANGELOG.md
-            fi
-        fi
-        if [[ "${line}" =~ "## v${chart_version}" ]] ; then
-            in_changelog=true
-        fi
-
-    done < ${chart_dir}/CHANGELOG.md
-    gh release create ${tag} --repo ${repo_full_name} --title "${title}" --notes-file /tmp/CHANGELOG.md
-    rm /tmp/CHANGELOG.md
-}
diff --git a/charts/scripts/create_gh_release.sh b/charts/scripts/create_gh_release.sh
deleted file mode 100755
index 598495f65..000000000
--- a/charts/scripts/create_gh_release.sh
+++ /dev/null
@@ -1,45 +0,0 @@
-#!/bin/bash
-
-set -e
-
-CURR_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
-[ -d "$CURR_DIR" ] || { echo "FATAL: no current dir (maybe running in zsh?)";  exit 1; }
-TOP_DIR=$CURR_DIR/..
-
-# shellcheck source=common.sh
-source "$CURR_DIR/common.sh"
-
-if [[ -z "${CHART_NAME}" ]] ; then
-    abort "Need to specify the chart you wish to publish"
-fi
-chart_dir="${TOP_DIR}/${CHART_NAME}"
-
-if [[ ! -d "${chart_dir}" ]] ; then
-    abort "${chart_dir} is not a directory"
-fi
-
-#########################################################################################
-if ! command -v gh 2> /dev/null ; then
-    info "gh doesn't exist, install before continuing"
-    exit 1
-fi
-if [[ ${CHART_VERSION:-} != v7.* ]]; then
-    abort "CHART_VERSION must be set to a 'v7.*' string"
-fi
-thisversion=${CHART_VERSION%v}
-chart_version=chart/v${thisversion}
-git fetch
-
-if ! git rev-parse ${chart_version} >/dev/null 2>&1 ; then
-    info "${chart_version} isnt a git tag, aborting"
-    exit 1
-fi
-
-if [[ $thisversion =~ ^[0-9]+\.[0-9]+\.[0-9]+$ ]] ; then
-    create_chart_release $thisversion $chart_dir
-else
-    info "${thisversion} doesnt look like a GA version, not creating release"
-fi
-
-
-exit 0
diff --git a/charts/scripts/image_tag_changelog_update.sh b/charts/scripts/image_tag_changelog_update.sh
index 8f8717b06..ec9d2d6fc 100755
--- a/charts/scripts/image_tag_changelog_update.sh
+++ b/charts/scripts/image_tag_changelog_update.sh
@@ -21,7 +21,7 @@ CHART_DIR="${TOP_DIR}/${CHART_NAME}"
 if [[ ${CHART_VERSION:-} != v7.* ]]; then
     abort "CHART_VERSION must be set to a 'v7.*' string"
 fi
-chart_version=${CHART_VERSION%v}
+chart_version=${CHART_VERSION#v}
 
 new_changelog=${CHART_DIR}/CHANGELOG.new.md
 ambassador_changelog_link="https://github.com/emissary-ingress/emissary/blob/master/CHANGELOG.md"
diff --git a/charts/scripts/push_chart.sh b/charts/scripts/push_chart.sh
index b9ebc2271..bf45d536b 100755
--- a/charts/scripts/push_chart.sh
+++ b/charts/scripts/push_chart.sh
@@ -29,7 +29,7 @@ fi
 if [[ ${CHART_VERSION:-} != v7.* ]]; then
     abort "CHART_VERSION must be set to a 'v7.*' string"
 fi
-thisversion=${CHART_VERSION%v}
+thisversion=${CHART_VERSION#v}
 
 repo_key=
 if [[ -n "${REPO_KEY}" ]] ; then
diff --git a/charts/scripts/update_chart_changelog.sh b/charts/scripts/update_chart_changelog.sh
index 8e7df087e..b2ce6cd8f 100755
--- a/charts/scripts/update_chart_changelog.sh
+++ b/charts/scripts/update_chart_changelog.sh
@@ -17,7 +17,7 @@ CHART_DIR="${TOP_DIR}/${CHART_NAME}"
 if [[ ${CHART_VERSION:-} != v7.* ]]; then
     abort "CHART_VERSION must be set to a 'v7.*' string"
 fi
-chart_version=${CHART_VERSION%v}
+chart_version=${CHART_VERSION#v}
 
 new_changelog=${CHART_DIR}/CHANGELOG.new.md
 rm ${new_changelog} > /dev/null 2>&1 || true
```

kflynn reviewed (approved):
👍  I think this hits everything I was asking for. 🙂
